test/e2e/etcd_failure.go

```diff
@@ -0,0 +1,171 @@
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/wait"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const etcdFailureDuration = 20 * time.Second
+
+type etcdFailureCase struct {
+	c  *client.Client
+	ns string
+}
+
+var _ = Describe("Etcd failure", func() {
+	defer GinkgoRecover()
+
+	var c *client.Client
+	var ns string
+
+	BeforeEach(func() {
+		By("preparing namespace")
+
+		ns = "e2e-test-" + string(util.NewUUID())
+		var err error
+
+		c, err = loadClient()
+		Expect(err).NotTo(HaveOccurred())
+
+		Logf("creating new namespace")
+		_, err = createNamespaceIfDoesNotExist(c, ns)
+		Expect(err).NotTo(HaveOccurred())
+
+		Logf("creating replication controller before failure")
+		Expect(RunRC(RCConfig{
+			Client:    c,
+			Name:      "baz",
+			Namespace: ns,
+			Image:     "nginx",
+			Replicas:  1,
+		})).NotTo(HaveOccurred())
+	})
+
+	It("should recover from network partition with master", func() {
+		var dropEtcdPackets sync.Once
+		testEtcdFailure(etcdFailureCase{c, ns}, func() {
+			dropEtcdPackets.Do(func() {
+				masterExec("sudo iptables -A INPUT -p tcp --destination-port 4001 -j DROP")
+			})
+		}, func() {
+			masterExec("sudo iptables -D INPUT -p tcp --destination-port 4001 -j DROP")
+		})
+	})
+
+	It("should recover from SIGKILL", func() {
+		testEtcdFailure(etcdFailureCase{c, ns}, func() {
+			masterExec("pgrep etcd | xargs -I {} sudo kill -9 {}")
+		}, func() { /* nothing */ })
+	})
+})
+
+func testEtcdFailure(efc etcdFailureCase, cause func(), fix func()) {
+	// This test requires SSH, so the provider check should be identical to
+	// there (the limiting factor is the implementation of util.go's
+	// getSigner(...)).
+	if !providerIs("gce", "gke") {
+		By(fmt.Sprintf("Skipping reboot test, which is not implemented for %s", testContext.Provider))
+		return
+	}
+
+	By("failing etcd")
+
+	stop := make(chan struct{}, 1)
+	go util.Until(func() {
+		defer GinkgoRecover()
+		cause()
+	}, 1*time.Second, stop)
+	time.Sleep(etcdFailureDuration)
+	stop <- struct{}{}
+	fix()
+
+	assertEtcdRecovers(efc)
+}
+
+func masterExec(cmd string) {
+	Logf("sshing into master %v to execute '%v'", getMasterHost(), cmd)
+	stdout, stderr, code, err := SSH(cmd, getMasterHost()+":22", testContext.Provider)
+	Expect(err).NotTo(HaveOccurred())
+	if code != 0 {
+		Failf("master exec command, '%v' failed with exitcode %v: \n\tstdout: %v\n\tstderr: %v", cmd, code, stdout, stderr)
+	}
+}
+
+func assertEtcdRecovers(efc etcdFailureCase) {
+	By("asserting etcd recovery")
+	pc := efc.c.Pods(efc.ns)
+	rcSelector := labels.Set{"name": "baz"}.AsSelector()
```

Comment:
The rest of this function seems like a more general test for replication controller recovery.  Can you factor this out into code that is re-used in that test and this one, rather than cutting and pasting?